rfc/006-backfill-blocks.md

```diff
@@ -87,15 +88,72 @@ changing consensus parameters, app specific needs, or a local config).
 This accommodates the use case of a truncated node that wishes to become an
 archive node. Note this could be done simultaneously alongside consensus.
 
-### Backfill verification
+### Verification
 
 Nodes will need to verify these prior blocks. This can be achieved by first
 retrieving the header at the base height from the block store. The node then
 checks that the `LastBlockID` corresponds with the hash calculated from the
-header in the new block directly below. The node can therefore trust the header.
-This then can be followed with using the hashes in the trusted header to
-validate the other block and state info and continue to sequentially verify the
-blocks below.  
+header in the new block directly below:
+
+```go
+header[height].LastBlockID == hash(header[height-1])
+```
+
+The node can therefore trust the new header. This then can be followed with
+using the hashes in the trusted header to validate the other block and state
+info. This is recursively done till the backfill height.  
+
+### Design
+
+The backfill mechanism would reside in the blockchain reactor. One may view it
+similarly to fast syncs position within the reactor as a service that can be
+started and stopped. Backfill, however, would not be an optional configuration.
+
+Two new messages would be added to allow for the passing of state data:
+
+```proto
+message StateDataRequest {
+  int64 height = 1;
+}
+```
+
+```proto
+message StateDataResponse {
+  int64 height = 1;
+  ValidatorSet validator_set = 2;
+  ConsensusParams consensus_params = 3;
+  ABCIResponses abci_responses = 4;
```

Comment:
Do we need ABCIResponses for previous blocks?

Reply:
It's part of the state store so for completeness we should include this.

There is however the proposal to remove it: https://github.com/tendermint/tendermint/issues/3533